packages/now-build-utils/src/detect-builders.ts

```diff
@@ -392,119 +556,409 @@ function validateFunctions(files: string[], { functions = {} }: Options) {
+      occurrences.map(name => `"${name}"`)
+    );
+
+    return {
+      defaultRoute: null,
+      dynamicRoute: null,
+      routeError: {
+        code: 'conflicting_file_path',
         message:
-          'Your `package.json` file is missing a `build` property inside the `scripts` property.' +
-          '\nMore details: https://zeit.co/docs/v2/platform/frequently-asked-questions#missing-build-script',
-      });
-      return { errors, warnings, builders: null };
+          `Two or more files have conflicting paths or names. ` +
+          `Please make sure path segments and filenames, without their extension, are unique. ` +
+          `The path "${fileName}" has conflicts with ${messagePaths}.`,
+      },
+    };
+  }
+
+  const out = createRouteFromPath(
+    fileName,
+    Boolean(options.featHandleMiss),
+    Boolean(options.cleanUrls)
+  );
+
+  return {
+    defaultRoute: out.route,
+    dynamicRoute: out.isDynamic ? out.route : null,
+    routeError: null,
+  };
+}
+
+// Checks if a placeholder with the same name is used
+// multiple times inside the same path
+function getConflictingSegment(filePath: string): string | null {
+  const segments = new Set<string>();
+
+  for (const segment of filePath.split('/')) {
+    const name = getSegmentName(segment);
+
+    if (name !== null && segments.has(name)) {
+      return name;
     }
 
-    // We allow a `public` directory
-    // when there are no build steps
-    const outDir = outputDirectory || 'public';
+    if (name) {
+      segments.add(name);
+    }
+  }
 
-    // If `outputDirectory` is an empty string,
-    // we'll default to the root directory.
-    if (hasDirectory(outDir, files) && outputDirectory !== '') {
-      frontendBuilder = {
-        use: '@now/static',
-        src: `${outDir}/**/*`,
-        config: {
-          zeroConfig: true,
-          outputDirectory: outDir,
-        },
-      };
-    } else if (
-      builders.length > 0 &&
-      files.some(f => !f.startsWith('api/') && f !== 'package.json')
-    ) {
-      // Everything besides the api directory
-      // and package.json can be served as static files
-      frontendBuilder = {
-        use: '@now/static',
-        src: '!{api/**,package.json}',
-        config: {
-          zeroConfig: true,
-        },
-      };
+  return null;
+}
+
+// Takes a filename or foldername, strips the extension
+// gets the part between the "[]" brackets.
+// It will return `null` if there are no brackets
+// and therefore no segment.
+function getSegmentName(segment: string): string | null {
+  const { name } = parsePath(segment);
+
+  if (name.startsWith('[') && name.endsWith(']')) {
+    return name.slice(1, -1);
+  }
+
+  return null;
+}
+
+// Counts how often a path occurs when all placeholders
+// got resolved, so we can check if they have conflicts
+function pathOccurrences(fileName: string, files: string[]): string[] {
+  const getAbsolutePath = (unresolvedPath: string): string => {
```

Comment:
Can you hoist `getAbsolutePath()` up into the top scope?


Also, we could probably call `getAbsolutePath()` once per file prior to calling `pathOccurrences()` or even `getApiRoute()` so it only happens N times instead of N^2 times.

Reply:
Basically, we need this inner loop to do as little as possible, and preferable no regex.

Reply:
@styfle good point, I've added a cache map so we can just get it from there or add it if it wasn't there yet.